oximeter/db/src/client.rs

```diff
@@ -642,34 +626,78 @@ impl DbWrite for Client {
+    // table. If that fails, those schema are _also_ removed from the internal
+    // cache.
+    //
+    // TODO-robustness There's still a race possible here. If two distinct clients receive new
+    // but conflicting schema, they will both try to insert those at some point into the schema
+    // tables. It's not clear how to handle this, since ClickHouse provides no transactions.
+    // This is unlikely to happen at this point, because the design is such that there will be
+    // a single `oximeter` instance, which has one client object, connected to a single
+    // ClickHouse server. But once we start replicating data, the window within which the race
+    // can occur is much larger, since it includes the time it takes ClickHouse to replicate
+    // data between nodes.
+    //
+    // NOTE: This is an issue even in the case where the schema don't conflict. Two clients may
+    // receive a sample with a new schema, and both would then try to insert that schema.
+    async fn flush_new_schema_or_remove(
+        &self,
+        new_schema: BTreeMap<TimeseriesName, String>,
+    ) -> Result<(), Error> {
         if !new_schema.is_empty() {
             debug!(
                 self.log,
                 "inserting {} new timeseries schema",
                 new_schema.len()
             );
-            let body = format!(
-                "INSERT INTO {db_name}.timeseries_schema FORMAT JSONEachRow\n{row_data}\n",
-                db_name = crate::DATABASE_NAME,
-                row_data = new_schema.join("\n"),
+            const APPROX_ROW_SIZE: usize = 64;
+            let mut body = String::with_capacity(
+                APPROX_ROW_SIZE + APPROX_ROW_SIZE * new_schema.len(),
             );
-            self.execute(body).await?;
+            body.push_str("INSERT INTO ");
+            body.push_str(crate::DATABASE_NAME);
+            body.push_str(".timeseries_schema FORMAT JSONEachRow\n");
+            for row_data in new_schema.values() {
+                body.push_str(row_data);
+                body.push_str("\n");
+            }
+
+            // Try to insert the schema.
+            //
+            // If this fails, be sure to remove the schema we've added from the
+            // internal cache. Since we check the internal cache first for
+            // schema, if we fail here but _don't_ remove the schema, we'll
+            // never end up inserting the schema, but we will insert samples.
+            if let Err(e) = self.execute(body).await {
+                debug!(
+                    self.log,
+                    "failed to insert new schema, removing from cache";
+                    "error" => ?e,
+                );
+                let mut schema = self.schema.lock().await;
+                for name in new_schema.keys() {
+                    schema
+                        .remove(name)
+                        .expect("New schema should have been cached");
+                }
```

Comment:
Is it correct to say that overzealously removing from the cache is low-risk, because the next time we try to do anything with a schema with a given name, we will check the DB for it anyway? I'm thinking about a scenario where there is an old value `V0` for `name` in `schema` and a new one for it `V1` in `new_schema`, we fail to insert `new_schema`, and then here we remove `(name, V0)` from `schema` even though it's not the one that failed to be inserted.

Reply:
I think it is low-risk, yeah. I considered just nuking the cache entirely, but that seems a bigger change than I wanted to fit in right before a release.

I'm not sure that the scenario you describe is possible though. If we get to the point of finding two schema for `name`, `V0` and `V1`, then they have to be equal or we ignore the sample in any case. So at this point in the code, we have exactly one schema in the cache, and at most one in the database. That sound right?

Reply:
Sounds good!